go/runtime/host/sgx/ecdsa.go

```diff
@@ -129,19 +164,33 @@ func (ec *teeStateECDSA) Update(ctx context.Context, sp *sgxProvisioner, conn pr
 
 	// Verify the quote so we can catch errors early (the runtime and later consensus layer will
 	// also do their own verification).
-	_, err = quote.Verify(quotePolicy, time.Now(), tcbBundle)
-	var tcbErr *pcs.TCBOutOfDateError
-	switch {
-	case err == nil:
-	case errors.As(err, &tcbErr):
-		sp.logger.Error("current TCB is not up to date",
-			"kind", tcbErr.Kind,
-			"tcb_status", tcbErr.Status.String(),
-			"advisory_ids", tcbErr.AdvisoryIDs,
-		)
-		return nil, tcbErr
-	default:
-		return nil, fmt.Errorf("quote verification failed: %w", err)
+	// Check bundles in order: fresh first, then cached, then try downloading again if there was
+	// no scheduled refresh this time.
+	var tcbBundle *pcs.TCBBundle
+	if err = ec.verifyBundle(quote, quotePolicy, freshTCBBundle, sp, "fresh"); err == nil {
+		tcbBundle = freshTCBBundle
+		tcbCache.cache(freshTCBBundle)
+	} else if err = ec.verifyBundle(quote, quotePolicy, cachedTCBBundle, sp, "cached"); err == nil {
+		tcbBundle = cachedTCBBundle
+	} else {
+		// If downloaded already, don't try again but just return the last error.
+		if freshTCBBundle != nil {
+			return nil, err
+		}
+
+		// If not downloaded yet this time round, try forcing. Any errors are fatal.
+		var downloadedTCBBundle *pcs.TCBBundle
+		if downloadedTCBBundle, err = ec.downloadBundle(ctx, sp, &quote); err != nil {
```

Comment:
Feels like we are repeating ourselves. Maybe we merge this part with the upper one.
```
	tcbBundle, err = func() (*pcs.TCBBundle, error) {
		if cachedTCBBundle, err := tcbCache.check(); err == nil {
			if err = ec.verifyBundle(quote, quotePolicy, freshTCBBundle, sp, "cached"); err == nil {
				return cachedTCBBundle, nil
			}
		}

		downloadedTCBBundle, err := ec.downloadBundle(ctx, sp, &quote);
		if err != nil {
			return nil, err
		}
		if err = ec.verifyBundle(quote, quotePolicy, freshTCBBundle, sp, "cached"); err != nil {
			return nil, err
		}
		tcbCache.cache(downloadedTCBBundle)
		return downloadedTCBBundle, nil
	}()
	if err != nil {
		return nil, err
	}
```

Reply:
I'm not sure the snippets are equivalent: the download must not be unconditional, but if it succeeds, its bundle should take precedence over whatever was in the cache.

Reply:
You are right. I thought that refresh is mandatory, but is not, as we use the cached bundle if download fails, which fine.